### ISSUE TITLE
Add quantified superclasses to Bifoldable and Bitraversable

### ISSUE THREAD

**Issue by Icelandjack** [haskell/core-libraries-committee#93]
Add qualified superclasses to `Bifoldable` and `Bitraversable`, obviously #91 takes precedence but I think this will not break much if anything.

```haskell
class (forall a. Foldable (bi a)) => Bifoldable bi

class (forall a. Traversable (bi a), Bifunctor bi, Bifoldable bi) => Bitraversable bi
```

Related proposals

* #91 Add quantified superclass to `Bifunctor`
* #92 Add `Profunctor` to base (with a quantified superclass)

**Comment by Bodigrim** [comment]
The proposal can use some explanation why this change is desirable.

**Comment by Bodigrim** [comment]
@Icelandjack @Topsii if anyone is still interested in this proposal, please flesh out some justification / motivation for the change.

**Comment by Bodigrim** [comment]
@Icelandjack @Topsii unless there is some progress within two weeks, I'll close as abandoned.

**Comment by Bodigrim** [comment]
Closing as abandoned, feel free to reopen when there is a renewed interest.

**Comment by phadej** [comment]
For the record, if someone wants to resurrect this proposal later.

current `base` (`4.17.0.0`) has `Bitraversable` (and `Bifoldable`) instances for tuples up to arity 7.

If this proposal is going to be implemented, if will force addition of (controversial) `Foldable` instances to these tuples.

**Comment by RyanGlScott** [comment]
@ekmett is currently in the middle of redesigning the `bifunctors` library as part of a general library overhaul, which will substantially change the API. One of the changes that was made during this overhaul was to change the `Functor` instances for `Biff` and `Tannen` to match what is proposed in https://github.com/haskell/core-libraries-committee/issues/93#issuecomment-1582781751.

In my opinion, there is no need to wait for the library overhaul to finish before we apply these `Functor` instance changes, especially since it is unclear when the overhaul will be complete. If the `Functor` instance changes are required to make the code compile with the new quantified superclasses for `Bifoldable`/`Bitraversable`, then that's an even stronger argument for making the `Functor` instance changes now.

**Comment by RyanGlScott** [comment]
> They are not required to make the code compile. The only thing in the bifunctors package that does not compile with the new quantified superclasses for Bifoldable/Bitraversable is the https://github.com/ekmett/bifunctors/pull/107.

Ah, my mistake. (Sorry for not noticing that issue sooner.)

> It seems wise to wait for the quantified superclass changes in base before changing the instances in bifunctors

Certainly, I am in no hurry to change anything. Once the changes land upstream in `base`, remind me and we can change `bifunctors` in corresponding ways.

**Comment by chshersh** [comment]
I'm not against the proposal. It feels like a right move to me. But I'm worried that the definition of `Bitraversable` becomes too unaccessible.

```haskell
class (forall a. Traversable (t a), Bifunctor t, Bifoldable t) => Bitraversable t where
```

I always like the fact that read `base` code without deep knowledge of Haskell. And maybe learn a few things from a couple of functions. But here, for the first time, I feel like `base` is becoming too complicated.

I don't want to block the proposal. Well, if things are complicated, it's life 🙂 

But I'd like to request some elaborations in the discussions on the signature. At least some links to the QuantifiedConstraints feature in GHC.

This feature is still quite new in `base`, so it would be nice to help readers to understand, thy this class looks different from others.

**Comment by mixphix** [comment]
If you're legitimately trying to use `Flip` as a `Bifunctor`, why are you using `Flip` to begin with?

I wouldn't even be opposed to the `Monoid` superclass of `Applicative`, frankly. These sorts of structural symmetries can and should be expressed in the type system. Making it harder to do silly things that can end up as footguns seems to be a good direction in which to steer.

**Comment by phadej** [comment]
> Or more fundamentally why don't we relate Applicative (aka MonoidalFunctor) with Monoid?

Because `[]`, `Maybe` and some other structures where `Applicative` structure is not the same as `Monoid` one. 

We could also want `forall a. Monoid (f a) => Alternative (f a)` 

But that also fails because `Maybe` and `IO`.

However, every `Bifunctor` is a functor by definition.

**Comment by ekmett** [comment]
>  And if the Functor/Foldable/Traversable instances in bifunctors had been defined with the weaker context intially, would this proposal even exist?!

The problem is they _aren't_ weaker contexts.

They would be _different_ contexts, requiring us to make even more distinctions between types for purely administrative plumbing reasons. Instead of 2 `Tannen` and `Cayley` constructions, we'd have 3, most other types would have to split as well to provide the user the choice between picking up a spurious `Functor (p a)` constraint that can't be discharged off of `Bifunctor p` or `Profunctor p`.

The original point of creating these superclasses was to get to a world where we could collapse more types together and actually make it so `Functor (p a)` is legitimately a weaker constraint than `Bifunctor p`.

The mildly imperfect nature of `Flip p`'s context hangs on the lack of a 'functor in the first of two arguments' class, which I used to call `PFunctor p` in some code in the past, a moral superclass of `Bifunctor` that could exist if we wanted to push further in that direction. The only real benefit would be for types like Flip, which seems to be a significantly more marginal improvement.

-----------

As for removing `first` or `second`, it is more like the 'Monad of No Return' proposal or removing `(/=)` from `Eq`. It seems like a plausible and pleasing step from a minimalist perspective, but it does seem prone to break code. There's a non-technical wibble I'd put on it is that it is a rather ugly asymmetry, but it puts the CLC in the position of forcing a style on the community for how instances are written, and for which there isn't a good answer. What do I mean?

Well, `second = fmap` and `second = bimap id` are both implementations that could be made top level default definitions for `second`, but you'll find the choice of which is used to be roughly equally distributed across actual instances. The former gets to ride the coattails of the free code GHC can derive for you. The latter leans on your possibly large and hand-written Bifunctor instance that you may not want to duplicate code for. If users had been using the latter and had simply defined `fmap = second`, then changing to `second = fmap` would break their existing code introducing non-termination in a GHC release silently. If users had been using the former you just made their code less efficient and possibly changed the asymptotic behavior by building up thunks of the form `(id x)` everywhere on the left hand side of their non-strict structures for each pass. This is the reason `(>>)`'s ghost lingers on inside `Monad`. We can't remove it without risking code introducing silent bottoms in perfectly legal code that would still typecheck. Admittedly, this is why we have warnings for non-canonical implementations, so we could gain the option to make such a cutover in the future, and since Functor is derivable in practice the pain is lessened.

Finally the removal of `second` prompts similar questions in `Arrow` and `Strong` from `profunctors` where such a removal is technically possible, but more similarly fraught with inefficiency and silently introduced non-termination wibbles. e.g. using Data.Tuple.swap you could define _Arrow_'s

```
second f = arr swap . Control.Arrow.first f . arr swap
```

and remove it from the class, but you risk the same termination issues in some user code above that had previously defined (***) in terms of `first` and `second`, and would introduce multiple walks through a structure that likely didn't need it before.

The case for _Strong_'s `second'` is
 
```
second' f = dimap swap swap . first'
```

where you'd risk introducing quiet (if not completely silent) non-termination on release in user code that had defined first' in terms of `second'` in similar way by only defining `second'` and using the existing MINIMAL pragma, and forces multiple passes over possibly deep structure.

So in all three cases I'd think long and hard and probably be opposed to their removal on possible silent-non-termination-introduction grounds alone.

**Comment by phadej** [comment]
@ekmett could you comment on the fact that there are instances where `second` and `fmap` have different laziness properties.

**Comment by mixphix** [comment]
+1

---

Since laziness is not a property that is captured by Haskell's type system, I think its behaviour ought to be considered an implementation detail in any case, and that we should not feel compelled to establish a complete description in full generality. The simplification to existing type signatures that this change will bring forth is already a drastic improvement to the current situation.

**Comment by angerman** [comment]
Strong -1 for me on this. With https://github.com/haskell-foundation/foundation/pull/572 being in recent memory (and that being one of the _public_ visible issues), I can not see myself supporting this.

This needs a clear deprecation/migration story. Maybe I missed it? This to me appears (so far), that the compiler including this change will simply reject existing code. This must stop. The compiler should *warn* about this change going to happen for _at least_ two versions (ideally with a link to the migration guide).

Irrespective of the content of this proposal, I can not support this due to:
- abrupt breaking change.
- no deprecation/change story.

**Comment by hasufell** [comment]
I can see myself potentially supporting this, but I don't really get the motivation. 

To me, adding superclass constraints is something fairly disruptive. That may still have benefits, but "my type signature will be shorter" is not one of those, IMO.

So I feel I can't really say whether the benefits will outweigh the potential costs, because I find them unclear.

----

Hence I abstain.

**Comment by chshersh** [comment]
+1

<hr/>

Although, I agree with @angerman's assessment about not having a proper deprecation mechanism for this proposal, I don't think it exists currently in GHC. Moving to QuantifiedConstraints sounds like a direction `base` has already taken, and I feel really weird drawing the line before `Bifoldable` and `Bitraversable`.

I like the new documentation additions, and the best we can do is to write a good migration guide.

Apart from potential future addition of `Profunctor` to `base` with a possible `QuantifiedConstraint`, I don't see any other changes w.r.t. to QuantifiedConstraints on the horizon. So we might as well just bite the bullet here (and I want to notice that the required impact assessment and patch contribution work has already been done). 

Note, it is a breaking change for some but it could be a desirable breaking change. As @Bodigrim pointed out, specifying `Traversable` constraints is quite awkward.

I do sympathise with issues similar to `foundation` but I agree that it's impossible to satisfy everyone. Part of the reason why the issue in the `foundation` happened in the first place is that the relationship between `Functor` and `Bifunctor` wasn't enforced by the compiler. Proceeding ahead with this proposal prevents potential points of confusion and misunderstanding how the provided API should be used in the future.

**Comment by parsonsmatt** [comment]
-1 for now, conditional on a GHC proposal that can satisfy some of the naysayers here.

---

I'm in favor of the change, but also sympathetic to @angerman on the potential for increasing the difficulty of upgrading code.

Moritz wants a way to *warn* users about these changes, so they can make changes and adjust their code in anticipation, *or* actively participtae in discussion that impacts them that they disagree with. I think this could be accomplished with a custom `WARNING` pragma on the class - given a proper name (ie `bifoldable-superclass-compat`) would allow people to silence the warning for their project once they've accounted for it, and could describe how to migrate, *and* where to comment if they feel that it's a bad thing. 

However, the `WARNING` pragma attached directly to the class - would that happen at every *use* site? Or just instance definition sites? I expect a WARNING at every *use* site would be a much more invasive and annoying problem. This points to a need to attach `WARNING` to *either* use site (ie `foo :: (Bifunctor f) => f a b -> f a b`) and definition site (ie `instance Bifunctor Either`).

I think I'd be OK with a `WARNING` at current tech levels provided that the message was sufficient, and I would not gate approval of this PR on a GHC feature that would allow use-site vs instance-definition-site specifictiy on warnings. 

I *would* like to see a more fully fleshed out proposal for providing helpful compiler messages - I suspect that a magic constraint, similar to `Unsatisfiable`, could be used - I'm imagining something like,

```haskell
type GoodMessage = 'Text "actually good message here, incl discussion link and migration guide"
type WarningCategory = "bifunctor-superclass-compat"

class
  (WarnIfMissing WarningCategory GoodMessage (forall a. (Functor (f a)))
  => Bifunctor f
```

where `WarnIfMissing` would search for the constraint and warn if it were unsatisfied. 

Right now, we simply don't have a mechanism to go from "This compiles with no warnings" to "This does not compile at all" for 'adding a superclass constraint'. We've gone through this a bunch as a community - `Applicative m => Monad m`, `Semigroup m => Monoid m`, etc. Every time we do this, it's a huge deal, and there are *more* superclasses we want to put in and haven't because of the nuisance.

---

The Foldable/Traversable proposal that generalized the functions in `Prelude` and `Data.List` was a hot topic as people were worried that code like `length (Just [1,2,3])` and `sum (1, 2)` would be confusing and lead to bugs, and the `instance {Functor,Foldable,Traversable} ((,) a)` came under scrutiny. Some folks didn't like it on aesthetic grounds, even if the instances are lawful and technically correct.

The `(forall a. Functor (f a)) => Bifunctor f` proposal rekindle those flames - if you really didn't like having tuples that could be folded and mapped over the right variable only, then you could "exit" by defining `data Tuple2 a b = Tuple2 a b`, *omit* the `Functor` instance, and *only* provide a `Bifunctor Tuple2` instance. That change removed the option for exit - now you were forced to provide a `Functor (Tuple2 a)` instance. 

*This* proposal takes the basic idea farther - now `Tuple2` must have a `Foldable` and `Traversable` instance for `Bifoldable` and `Bitraversable` to exist.

In @vincenthz's words,

> as some previously captured here, the whole point of not having Functor was to not have biased implementation available to the user.
> 
> This is hugely disappointing again from the ghc / base maintainers to completely change the semantics of a class after multiple major releases, and fly in the face of stability and backwards compatibility that are required to have a decent ecosystem and language.

Vincent wants a way to forbid users from having a biased implementation - but the concrete problem trying to solve is the potential for *confusion* - expecting that `fmap (+1) (1, 2) == (2, 3)` and then getting bugs when it really equals `(1, 3)`,  and *silent changes in semantics* -  a code change `(mkThing :: IO [a]) >>= pure . length` into `(mkThing :: IO (b, [a])) >>= pure . length` and again causing a problem.

This suggests a need for *use site specific warnings for type class instances* with custom names such that we can promote them to `-Werror`. "Hey, you're doing `fmap` on a tuple, which only maps over the second element. Consider `bimap` instead to explicitly target the element you want to map over." 

Something like [Custom Type Warnings (GHC proposal 454](https://github.com/ghc-proposals/ghc-proposals/pull/454)https://github.com/ghc-proposals/ghc-proposals/pull/454) could go a long way here.

**Comment by tomjaguarpaw** [comment]
-1

----

I was originally against https://github.com/haskell/core-libraries-committee/issues/91, then was persuaded by some reasonable arguments to change my mind, but in retrospect I'm not certain that was the correct decision. However, that proposal was about `Bifunctor`, a class that although I don't love it[^1],  it seems only overly abstruse rather than actively harmful.  This proposal is (in part) about `Bifoldable`. My belief is that `Foldable` is a terrible and harmful abstraction (it's impossible to use it for genuine generic programming, it can only be used for name overloading) and the `Bi` version of it is a just an ad hoc hack laid on top.

Unfortunately I think this proposal is a victim of a too-large base. `Bifunctor`/`foldable`/`traversable` don't belong in `base` at all, in my opinion. They should be in a separate package where it's easier to evolve them.

[^1]: are we going to have `Tri`/`Quadrafunctor`? Where does the madness stop?

**Comment by tomjaguarpaw** [comment]
> This suggests a need for use site specific warnings for type class instances

FWIW [stan](https://github.com/kowainik/stan/) can do this (and already has a built in rule for [`Foldable` methods on tuples](https://github.com/kowainik/stan/blob/72dda7c484ceb5313d561560ada5b05f7db4de63/src/Stan/Inspection/AntiPattern.hs#L212-L221)).

**Comment by parsonsmatt** [comment]
> > This suggests a need for use site specific warnings for type class instances
> 
> FWIW [stan](https://github.com/kowainik/stan/) can do this (and already has a built in rule for [`Foldable` methods on tuples](https://github.com/kowainik/stan/blob/72dda7c484ceb5313d561560ada5b05f7db4de63/src/Stan/Inspection/AntiPattern.hs#L212-L221)).

The last release of `stan` was around 3 years ago. It's an optional tool that isn't used often and doesn't support recent GHC versions. These facts make me feel like it's *more* important to have GHC support for this functionality.

**Comment by tomjaguarpaw** [comment]
Sure, the stan maintainers have chosen not to apply [patches bringing it more up-to-date](https://github.com/kowainik/stan/issues/423#issuecomment-1264348102). But I think it's important to note that in principle the information you seek is available through the GHC API. It doesn't necessarily need to be exposed as a warning/error through GHC itself.

**Comment by Bodigrim** [comment]
There are 3 votes in favor, 3 against, and 1 abstention. It means that there is no majority in support of the proposal.

I imagine one can raise a GHC proposal for a gradual migration, similar to what was done to `Semigroup` / `Monoid`: start with throwing a warning if there is `Bitraversable`, but no `Traversable`, over the course of several GHC versions, and make `Traversable` a superclass of `Bitravesable` only afterwards. 

@tomjaguarpaw @parsonsmatt @angerman @hasufell would the proposal have your support in case of gradual migration?

**Comment by hasufell** [comment]
I can't deny or confirm. My main issue is the motivation and the cost-benefit analysis. Both are unclear to me, so a follow up proposal should make this more clear: what can we implement that we could not before? Is this just convenience or "morally correct"?

**Comment by tomjaguarpaw** [comment]
I have the same point of view as @hasufell

**Comment by ekmett** [comment]
@hasufell It makes it so that I'm not perpetually lying in my constraints about what is required for something to be an instance of Bifunctor or Profunctor or the like, and I'm not continually forced to make unforced choices between constraints for instances on types. The status quo leaves me forced to make multiple completely equivalent types that only differ in which completely unnecessary constraints are required for their instances. This adds a perpetual burden on users to keep track of those useless distinctions. 

I believe strongly in writing types down and only requiring exactly what is required for each instance.

> what can we implement that we could not before? 

It is more what I _don't_ have to implement. It lets us collapse a number of types down into one, erasing unnecessary unforced distinctions.

If that isn't enough, do what you want.

**Comment by Bodigrim** [comment]
@ekmett to be clear: this proposal is about `Bifoldable` / `Bitraversable`. In contrast to them `Functor` is already a superclass of `Bifunctor`, which seems sufficient for your purposes?..

**Comment by gbaz** [comment]
> Bifunctor/foldable/traversable don't belong in base at all, in my opinion. They should be in a separate package where it's easier to evolve them

Sorry for going a bit off-topic here. The arguments about foldable have been hashed out enough at this point, but I really do want to speak up in terms of traversable! Its a good chunk of what is motivated in the original applicative papers, and for good reason. They go together great, and furthermore at this point the laws concerning traversable have been worked out in great detail, and turn out to be very elegant. There's a saying I've seen among some programmers that "its always `traverse`" meaning that its shocking how many type-tetris functions that are called for which can be accomplished with `traverse` instantiated at the right types. Its one of the best swiss-army knives in a Haskell programmers toolkit, and it absolutely has more than proven its weight as belonging in base.

I agree foldable is fairly messy, and is the result of a lot of irritating tradeoffs between efficiency and abstraction, and in that sense it is unfortunate that it is a superclass of traversable. But traversable on its own is extemely elegant, and we'd be much the poorer if it wasn't pervasively available without need for any dependencies.

**Comment by tomjaguarpaw** [comment]
> I really do want to speak up in terms of traversable! 

Sorry, I was unclear. When I said "`Bifunctor`/`foldable`/`traversable`" I intended the `Bi` to distribute, so I meant `Bifunctor`/`Bifoldable`/`Bitraversable` should be in a separate package.

I'm a very big fan of `Traversable` (even though I believe it's superseded by `Traversal`s). I'm an opponent of `Foldable`, but I'd never dare to say it belongs outside of `base`!

**Comment by ekmett** [comment]
@Bodigrim 

> to be clear: this proposal is about `Bifoldable` / `Bitraversable`. In contrast to them `Functor` is already a superclass of `Bifunctor`, which seems sufficient for your purposes?..

It is sufficient... right up until you to go to define the `Foldable`/`Traversable` instance for `Cayley`/`Tannen`.

Is it `(Foldable f, Bifoldable p) =>` or `(Foldable (p a), Foldable f) =>`?

Neither subsumes the other without the superclass constraint, forcing you into the dilemma of having to supply both options because someone somewhere is hating on `Foldable` and I meanwhile I just want to be a good ecosystem citizen providing solid reusable canonical types for things and to get on with my life. All you do is delay the unforced distinction if you don't just finish the job.

I find myself nodding along with @chshersh above that I kind of wish `Bifunctor` hadn't been folded into `base`. If it hadn't, it would still be in my `bifunctors` package, and I could have fixed this unilaterally, years ago.

As it is, I have code I don't even really remember that has been rotting in a repo for a year that builds on this, and will be rotting in a repo for another year until after this change is finally made, to have a compatible `base` version or forever if it is not.

**Comment by tomjaguarpaw** [comment]
I am in favour of gradual migration in any case, but @ekmett's comment helps me understand why this is actually desirable. If we're going to gradually migrate we should definitely start gradually migrating *now* so we don't forever get stuck in the limbo of "yeah we know we're going to do this but we're not going to do it yet".

**Comment by tomjaguarpaw** [comment]
^ @ekmett

**Comment by angerman** [comment]
I am _not_ fundamentally against evolution. I am fundamentally against abrupt evolution. The compiler _must_ accept code it accepted in the previous version, _unless_ it did warn (while still accepting) about upcoming changes for _at least_ two major versions.

**if** we have a proper migration story here, I won't be against it conceptually. But this needs to address the following:
- ensure existing code works for a generous grace period. IT MUST NOT STOP ACCEPTING CODE THAT CURRENTLY COMPILES WITHOUT ANY WARNINGS.
- ensure that the compiler gives good and helpful hints towards how to migrate (maybe point to something similar like the GHC Errors; ideally with a way to also provide feedback on that upcoming change. If a large enough subset of the community end up against that change, it must be reconsidered).

So I guess my stance is the same as @parsonsmatt's.

**Comment by Bodigrim** [comment]
Thanks all for the discussion. 

@Topsii I'm sorry, but as it stands the proposal does not command the majority of the committee. That said, judging from the comments, CLC members might think more favorable of the change, if there was a GHC proposal first, introducing warnings for `Bifoldable` without `Foldable` and `Bitraversable` without `Traversable` similar to the migration story for Applicative-Monad proposal with `-f[no-]warn-amp` flag in GHC 7.8.

There are also two auxiliary proposals here, shadowed by the main one:
* Laws for `Bifoldable a` when there is `Foldable a`,
* Instances of `Foldable` and `Traverable` for tuples. 

I'll raise them as separate issues to discuss.
